ruby/phase-1.md

```diff
@@ -0,0 +1,616 @@
+processes. We have agreed that this part should be left out of the Ruby SDK and handled by
+developers themselves. Potentially this functionality can be added as a separate Ruby gem.
+
+
+## Activities
+
+Each activity is going to be represented by its own class. This decision comes with a few benefits:
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
+- It plays nicely with Ruby's OOP model
+- It allows for future extensibility including 3rd party extensions
+- Activity can be easily tested by injecting a test context into the class instead of the real one
+
+By default we'll use the name of the class as activity name (when communicating with the server).
+This is of course overridable in case the SDK users would like to use their own naming schema. And
+we'll use the method `#execute` as the primary entrypoint for executing user code.
+
+```ruby
+class MyActivity < Temporal::Activity
+  activity 'my-activity' # custom name
+
+  def execute(arg_1, arg_2: nil)
+    loop do
+      sleep 1
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::CancellationRequested
+    activity.logger.info('💀')
+    # Accept cancellation request
+    raise Temporal::Error::ActivityCancelled, 'cancellation details here'
+  end
+end
+```
+
+Notes:
+
+- A custom activity name can be specified using `activity 'my-activity'` call (we can't use name,
+  because it is reserved to the actual class name in Ruby)
+- `activity` within `#execute` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- In case of a cancellation request `Temporal::Error::CancellationRequested` will get raised, which
+  can be accepted by raising `Temporal::Error::ActivityCancelled`
+
+```ruby
+class Temporal::Activity::Context
+  # Access to a pre-configured logger
+  def logger() -> Temporal::Logger
+
+  # Heartbeat will raise if a cancellation was requested
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+### Return values and errors
+
+An activity can communicate its result to the workflow by returning a value or raising an error. The
+return values will get serialized as a `Payload` object similar to activity input arguments.
+Similarly, an error will get serialized and re-raised from within the workflow (if an activity
+called "synchronously").
```

Comment:
Temporal requires that any application error is serialized to an instance of a Failure with failure_info set to [ApplicaitonFailure](https://github.com/temporalio/api/blob/2ed449fbdfedecd91d23b81fcc423f4665b3b4fe/temporal/api/failure/v1/message.proto#L87).

This rules out the design that serializes/deserializes Ruby exceptions. Another limitation of the proposed design that it is not going to work across languages. What would be the type to deseriaze PaymentMethodNotFound in Go, for example?

I think Ruby should follow the approach that Typescript and other SDKs took. See [Failure classes reference.](https://docs.temporal.io/docs/typescript/handling-failure#failure-classes-reference)

Reply:
Why do you think this is a problem? Go will just deserialize it as a generic Activity/Workflow error with a type set on a string attribute.

However I see many downsides of not following this approach — for example you won't be able to use Ruby's error matching when catching an exceptions (as these all would have to be generic ApplicationFailures).

So instead of something like this:

```ruby
begin
  PaymentActivity.execute(payment_method)
rescue PaymentActivity::PaymentMethodNotFound => e
  EmailActivity.execute('payment_failure', e.message)
end
```

You would have to do this every time you handle an error:

```ruby
begin
  PaymentActivity.execute(payment_method)
rescue Temporal::WorkflowError => e
  if e.type == 'PaymentMethodNotFound'
    EmailActivity.execute('payment_failure', e.message)
  else
    raise e # re-raise
  end
end
```

Reply:
Actually, in your second snippet there, it'd be `Temporal::ActivityError` since it was thrown by the activity. But that doesn't change the argument.

In most languages, wrapping an exception to know where it came from is very reasonable. If a user wants to throw an error out of some helper a workflow calls and then catch it, no prob. But when Temporal throws an error, it is nice to say all errors from an activity are activity errors, all errors from a workflow are workflow errors, etc. Users can unpack, but that's better than inconsistent wrapping.

Also, what if you didn't catch this error? That would throw out of the workflow as though it was thrown in the workflow. This may not seem important now, but is very important on things like cancellation. There is a difference between activity cancellation and workflow cancellation though they use the same type. But luckily we consistently wrap errors to remove ambiguity.

Reply:
@cretz Yep, my typo, meant `Temporal::ActivityError`.

Another thing you can do in Ruby is you can catch the superclass of the error. So if SDK users subclass their errors from `Temporal::ActivityError` they can catch that instead of any specific individual error.

As far as unhandled errors go — it gets raised in the workflow context which is equivalent of raising directly from a workflow, resulting in failing it. For workflow cancellations we can either raise a special SDK exception `Temporal::CancelWorkflow` or wrap it in a call on the context object — `workflow.cancel!`.

What is very important here is getting on the same page with regards to how writing a workflow should feel to the SDK users. Should the experience be of writing in-process code that gets distributed automagically (think local method/class calls) or should the user be constantly aware that they're dealing with a distributed system here (so instead of local calls it's more like RPC calls)? Examples:

1. In-process feel:

```ruby
class PaymentWorkflow < Temporal::Workflow
  def execute(amount, user_id, payment_method_id)
    payment = CreatePaymentActivity.execute!(amount, user_id, payment_method_id)

    payment_receipt = ProcessPaymentActivity.execute!(payment.id)

    EmailSuccessActivity.execute!(user_id, payment_receipt)
  rescue ProcessPaymentActivity::InsufficientFunds
    EmailFailureActivity.execute!(user_id, "You don't have enough funds")
  end
end
```

2. RPC feel:

```ruby
class PaymentWorkflow < Temporal::Workflow
  def execute(amount, user_id, payment_method_id)
    result = workflow.execute_activity(CreatePaymentActivity, arguments: [amount, user_id, payment_method_id])

    workflow.fail!(result.error) if result.failure?
    
    result = workflow.execute_activity(ProcessPaymentActivity, arguments: [payment.id])
    
    if result.failure? 
      if result.error.name == 'ProcessPaymentActivity::InsufficientFunds'
        workflow.execute_activity(EmailFailureActivity, arguments: [user_id, "You don't have enough funds"])
        workflow.complete!
      else
        workflow.fail!(result.error)
      end
    else
      workflow.execute_activity(EmailSuccessActivity, user_id, result.value)
    end

    workflow.complete!
  end
end
```

The main difference here is that in the 2nd example it feels like you're talking to a foreign service with more processes around it and less native flow control constructs — return values and errors get wrapped, each invocation feels like an RPC, much more noticeable SDK presence.

Please note that we're NOT discussing the workflow API right now. However it's good to start thinking about how we would want the SDK to feel and then put this as a proposal requirement (right now the process is a bit more backwards — we're trying to agree on a specific implementation first and requirements are a second thought).

Reply:
I think the first in-process feel is fine, but you would rescue a general temporal activity error. We don't implicitly unwrap it on purpose - so they can differentiate errors thrown by Temporal (i.e. from Temporal server and a completely separate worker) compared to some local method they called.

Not only that, but errors are not always serializable. Most SDKs can't really build that `ProcessPaymentActivity::InsufficientFunds` error.

So yes, it might look like:

```ruby
  def execute(amount, user_id, payment_method_id)
    payment = CreatePaymentActivity.execute!(amount, user_id, payment_method_id)

    payment_receipt = ProcessPaymentActivity.execute!(payment.id)

    EmailSuccessActivity.execute!(user_id, payment_receipt)
  rescue Temporal::ActivityError => err
    if err.cause.is_a?(Temporal::ApplicationError) and err.cause.type == 'InsufficientFunds'
      EmailFailureActivity.execute!(user_id, "You don't have enough funds")
    else
      raise
    end
  end
```

You can't do much better without me being able to point out a bunch of ambiguities. All activities in Temporal throw the same errors. This is not a Ruby specific thing. That Ruby or any other language may not have the cleanest way to handle nested exceptions (and none of them really do) I don't think is justification for trying to re-serialize Ruby-specific exceptions, or have Ruby not always throw what the server tells it to throw, etc. The server gives the failure, we don't unwrap it.